src/webnn/native/dml/GraphDML.cpp

```diff
@@ -1675,7 +1701,289 @@ namespace webnn::native ::dml {
     }
 
     MaybeError Graph::AddGru(const op::Gru* gru) {
-        return DAWN_UNIMPLEMENTED_ERROR("Gru hasn't been supported on DirectML.");
+        const auto inputsOperand = gru->Inputs();
+        DAWN_ASSERT(inputsOperand.size() >= 3 && inputsOperand.size() <= 6);
+        DAWN_ASSERT(mGraphEdgesMap.find(inputsOperand[0].Get()) != mGraphEdgesMap.end());
+        DAWN_ASSERT(mGraphEdgesMap.find(inputsOperand[1].Get()) != mGraphEdgesMap.end());
+        DAWN_ASSERT(mGraphEdgesMap.find(inputsOperand[2].Get()) != mGraphEdgesMap.end());
+        std::vector<std::shared_ptr<EdgeInfoBase>> inputEdges;
+
+        // Input: 4D tensor with the Sizes of { 1, seq_length, batch_size, input_size }.
+        // Need to reshape input from WebNN 3-D to DML 4-D.
+        auto inputEdge = mGraphEdgesMap[inputsOperand[0].Get()];
+        auto webnnInputDims = ConvertDimensions(inputsOperand[0].Get()->Shape());
+        std::vector<UINT> inputDims = {1, webnnInputDims[0], webnnInputDims[1], webnnInputDims[2]};
+        std::shared_ptr<DmlTensorDesc> inputDmlTensorDesc(new DmlTensorDesc);
+        if (!CreateDmlTensorDesc(mDmlTensorsDesc, inputDmlTensorDesc, &inputEdge->outputTensorDESC,
+                                 inputDims)) {
+            return DAWN_INTERNAL_ERROR("Failed to create DML tensor description.");
+        }
+        DML_TENSOR_DESC inputTensorDesc = {DML_TENSOR_TYPE_BUFFER, &inputDmlTensorDesc->bufferDesc};
+        inputEdges.push_back(inputEdge);
+
+        // Weight: 4D tensor with the Sizes of { 1, num_directions, 3 * hidden_size, input_size }.
+        // Need to reshape weight from WebNN 3-D to DML 4-D.
+        // Need to convert tensor flag to NONE.
+        ComPtr<IDMLOperator> dmlOperator;
+        auto constantWeightEdge = mGraphEdgesMap[inputsOperand[1].Get()];
+        auto webnnWeightDims = ConvertDimensions(inputsOperand[1].Get()->Shape());
+        std::vector<UINT> weightDims = {1, webnnWeightDims[0], webnnWeightDims[1],
+                                        webnnWeightDims[2]};
+        std::shared_ptr<DmlTensorDesc> constantWeightDmlTensorDesc(new DmlTensorDesc);
```

Comment:
Please add some annotations to explain why we need append identity for Gru's inputs.

Reply:
Optimized the code and annotations, PTAL thanks a lot!